From c324120c80e7e1b1897d2a8116da07f8cf6d4942 Mon Sep 17 00:00:00 2001
From: Wangda Tan <wangda@apache.org>
Date: Tue, 8 Dec 2015 14:25:16 -0800
Subject: [PATCH 1257/2848] YARN-4424. Fix deadlock in RMAppImpl. (Jian he via
 wangda)

(cherry picked from commit 7e4715186d31ac889fba26d453feedcebb11fc70)
(cherry picked from commit 7013f9d6cda88e72a839b1c55757615b55101beb)

Change-Id: I93b91be88361b62ac9ebe50d10a70f293bce7a0f
---
 .../server/resourcemanager/rmapp/RMAppImpl.java    |   22 +++++++-------------
 1 file changed, 8 insertions(+), 14 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmapp/RMAppImpl.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmapp/RMAppImpl.java
index 06d67e6..29fc2cb 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmapp/RMAppImpl.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmapp/RMAppImpl.java
@@ -495,21 +495,15 @@ public ApplicationSubmissionContext getApplicationSubmissionContext() {
 
   @Override
   public FinalApplicationStatus getFinalApplicationStatus() {
-    this.readLock.lock();
-    try {
-      // finish state is obtained based on the state machine's current state 
-      // as a fall-back in case the application has not been unregistered 
-      // ( or if the app never unregistered itself )
-      // when the report is requested
-      if (currentAttempt != null 
-          && currentAttempt.getFinalApplicationStatus() != null) {
-        return currentAttempt.getFinalApplicationStatus();   
-      }
-      return 
-          createFinalApplicationStatus(this.stateMachine.getCurrentState());
-    } finally {
-      this.readLock.unlock();
+    // finish state is obtained based on the state machine's current state
+    // as a fall-back in case the application has not been unregistered
+    // ( or if the app never unregistered itself )
+    // when the report is requested
+    if (currentAttempt != null
+        && currentAttempt.getFinalApplicationStatus() != null) {
+      return currentAttempt.getFinalApplicationStatus();
     }
+    return createFinalApplicationStatus(this.stateMachine.getCurrentState());
   }
 
   @Override
-- 
1.7.9.5

